Sources/FoundationEssentials/FileManager/FileManager+Files.swift

```diff
@@ -553,49 +553,61 @@ extension _FileManagerImpl {
                             ? FileAttributeType.typeDirectory
                             : FileAttributeType.typeRegular
                 case FILE_TYPE_PIPE: FileAttributeType.typeSocket
                 case FILE_TYPE_UNKNOWN: FileAttributeType.typeUnknown
                 default: FileAttributeType.typeUnknown
             }
 
-            let size: UInt64 = (UInt64(faAttributes.nFileSizeHigh) << 32) | UInt64(faAttributes.nFileSizeLow)
-            let creation: Date = Date(timeIntervalSince1970: faAttributes.ftCreationTime.timeIntervalSince1970)
-            let modification: Date = Date(timeIntervalSince1970: faAttributes.ftLastWriteTime.timeIntervalSince1970)
-            return [
+            let systemNumber = UInt64(info.dwVolumeSerialNumber)
+            let systemFileNumber = UInt64(info.nFileIndexHigh << 32) | UInt64(info.nFileIndexLow)
+            let referenceCount = UInt64(info.nNumberOfLinks)
+
+            let isReadOnly = info.dwFileAttributes & FILE_ATTRIBUTE_READONLY != 0
+            var posixPermissions = UInt16(_S_IEXEC | _S_IREAD)
+            if !isReadOnly {
+                posixPermissions |= UInt16(_S_IWRITE)
+            }
+
+            let size: UInt64 = (UInt64(info.nFileSizeHigh) << 32) | UInt64(info.nFileSizeLow)
+            let creation: Date = Date(timeIntervalSince1970: info.ftCreationTime.timeIntervalSince1970)
+            let modification: Date = Date(timeIntervalSince1970: info.ftLastWriteTime.timeIntervalSince1970)
+            var result: [FileAttributeKey : Any] = [
                 .size: _writeFileAttributePrimitive(size, as: UInt.self),
                 .modificationDate: modification,
                 .creationDate: creation,
                 .type: fatType,
+                .systemNumber: _writeFileAttributePrimitive(systemNumber, as: UInt.self),
+                .systemFileNumber: _writeFileAttributePrimitive(systemFileNumber, as: UInt.self),
+                .posixPermissions: _writeFileAttributePrimitive(posixPermissions, as: UInt.self),
+                .referenceCount: _writeFileAttributePrimitive(referenceCount, as: UInt.self),
 
-                // TODO(compnerd) support these attributes, remapping the Windows semantics...
-                // .posixPermissions: ...,
-                // .referenceCount: ...,
-                // .systemNumber: ...,
-                // .systemFileNumber: ...,
-                // .ownerAccountID: ...,
-                // .groupownerAccountID: ...,
-                // .ownerAccountName: ...,
-                // .groupOwnerAccountName: ...,
-                // .deviceIdentifier: ...,
+                // Uid is always 0 on Windows systems
+                .ownerAccountID: _writeFileAttributePrimitive(0, as: UInt.self),
+
+                // Group id is always 0 on Windows
+                .groupOwnerAccountID: _writeFileAttributePrimitive(0, as: UInt.self)
             ]
+
+            if fatType == .typeCharacterSpecial || fatType == .typeBlockSpecial {
```

Comment:
This seems ... inverted? If it is a special file then we add the device identifier? Shouldn't this be if the file is a disk file?

Reply:
I thought `st_rdev` (which is the field we're emulating here) was only relevant if it's a device node. I'm not actually sure what `dwVolumeSerialNumber` would be for e.g. `LPT` or `COM1`, mind; the documentation doesn't really say what it'll give for those. I'm not sure `dwVolumeSerialNumber` is the right value here, mind.

Reply:
Oh interesting, so this comes from what we have in SCL-F today and I missed a detail. SCL-F has:

```swift
statInfo.st_rdev = _dev_t(info.dwVolumeSerialNumber)
```

and then

```swift
#if os(Windows)
        result[.deviceIdentifier] = NSNumber(value: UInt64(s.st_rdev))
#endif

if type == .typeBlockSpecial || type == .typeCharacterSpecial {
    result[.deviceIdentifier] = NSNumber(value: UInt64(s.st_rdev))
}
```

So it seems that we only set `deviceIdentifier` for blockSpecial/characterSpecial on other platforms but on Windows we unconditionally set it. Should I go with that behavior here to mirror what SCL-F did?

Reply:
It looks like the tools-support-core repo doesn't use that value, so that might be a viable option

Reply:
I went ahead and updated this to a TODO, based on some GitHub searching I don't think this should affect anything in the toolchain build